astropy/cosmology/io.py

```diff
@@ -0,0 +1,351 @@
+
+        # parse file
+        mapping = json.loads(data)
+
+        return self.from_mapping(mapping, key=key)
+
+    def from_mapping(self, mapping, key=None):
+        """Load `~astropy.cosmology.Cosmology` from mapping object.
+
+        Parameters
+        ----------
+        mapping : mapping
+        key : hashable or None, optional
+            Any valid key into `mapping`.
+            If not None, this is used to select a cosmology if mapping is a
+            set of cosmologies.
+
+        Returns
+        -------
+        `~astropy.cosmology.Cosmology`
+
+        """
+        # get params subset from mpping
+        params = mapping[key] if key is not None else mapping
+
+        # determine class
+        subclasses = {c.__qualname__: c for c in
+                      self._cls.__subclasses__(deep=True)}
+        cls_name = params.pop("cosmology", self._cls.__qualname__)
+        try:
+            cosmo_cls = subclasses[cls_name]
+        except KeyError:
+            raise KeyError(f"{cls_name} is not a subclass of {cls.__qualname__}.")
+
+        # match keys to arguments
+        inspect_params = cosmo_cls._init_signature.parameters.values()
+        expected_params = set(
+            [
+                p.name
+                for p in inspect_params
+                if not ((p.kind == _VAR_POSITIONAL) or (p.kind == _VAR_KEYWORD))
+            ]
+        )
+        HAVE_KWARGS = any([p.kind == _VAR_KEYWORD for p in inspect_params])
+
+        # move non-parameters to metadata, if no **kwargs in signature
+        meta = params.pop("meta", None) or {}
+        for k, v in tuple(params.items()):
+            if k not in expected_params and not HAVE_KWARGS:
+                v = params.pop(k)  # make sure not in params
+                # TODO! raise an error / warning if repeat k?
+                meta.setdefault(k, v)
+
+            # TODO! parse list str when Quantity supports it
+            # s = repr([2, 3, 4] * u.kpc)[10:-1]
+            # num, unit = s.split("]")
+            # eval(num + "]") * u.Unit(unit)
+        meta = meta if meta else None  # empty meta -> None
+
+        # make cosmo instance
+        ba = cosmo_cls._init_signature.bind_partial(**params, meta=meta)
+        cosmo = cosmo_cls(*ba.args, **ba.kwargs)
+        return cosmo
+
+    def from_table(self, table, index=None):
+        """Instantiate a `~astropy.cosmology.Cosmology` from a |QTable|.
+
+        Parameters
+        ----------
+        table : `~astropy.QTable`
+        index : int or None, optional
+            The row from table.
+            TODO! allow for index to be a string -- the name fromo the row.
+
+        Returns
+        -------
+        `~astropy.cosmology.Cosmology`
+
+        """
+        # get 1-row of table
+        row = table.iloc[index] if index is not None else table
+        if len(row) != 1:  # check that it's one row
+            raise IncompatibleShapeError("Table must be 1-D.")
+
+        # special values
```

Comment:
I might be missing something, but I feel like the rest of the code here could be:
```
mapping = dict(row)  # Requires `meta` is in `row`, see below
return self.from_mapping(mapping)
```
There are a lot of options that appear to be supported in the code below. Are they *really* needed? If so, then you need to carefully document all the possibilities. Instead you can just say that the table columns need to map exactly onto the cosmology parameters. 

The issue of handling `meta` is a little tricky. It might be that keeping `meta` as an object column in the table is more sensible, even in the ECSV serialization. Since ECSV can now handle an object column, this could work well.

Reply:
I think this is resolved in the refactor. ``Cosmology.meta`` is held in ``Table.meta``. When Cosmology permits array inputs I would like to implement a row-specific meta. Perhaps the object column would be of use here, but I think that would be in addition to the Table meta. The row-specific meta would update/override that more general meta.